server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.script.field;
+
+import org.elasticsearch.common.util.Maps;
+
+import java.util.Map;
+import java.util.Objects;
+
+/** Metadata storage backed by a Map for compatibility with the ctx Map used by ingest and update scripts. */
+public class MapBackedMetadata {
+    public static final String INDEX = "_index";
+    public static final String ID = "_id";
+    public static final String VERSION = "_version";
+    public static final String ROUTING = "_routing";
+    public static final String SOURCE = "_source";
+
+    protected Map<String, Object> map;
+
+    public MapBackedMetadata(Map<String, Object> map) {
+        this.map = map;
+    }
+
+    public MapBackedMetadata(int size) {
+        this.map = Maps.newMapWithExpectedSize(size);
+    }
+
+    public String getIndex() {
+        return getString(INDEX);
+    }
+
+    public MapBackedMetadata setIndex(String index) {
+        return set(INDEX, index);
+    }
+
+    public String getId() {
+        return getString(ID);
+    }
+
+    public MapBackedMetadata setId(String id) {
+        return set(ID, id);
+    }
+
+    public Long getVersion() {
+        Object obj = getRawVersion();
+        if (obj == null) {
+            return null;
+        } else if (obj instanceof Number number) {
+            long version = number.longValue();
+            if (number.doubleValue() != version) {
+                // did we round?
+                throw new IllegalArgumentException(
+                    "version may only be set to an int or a long but was [" + number + "] with type [" + obj.getClass().getName() + "]"
+                );
+            }
+            return version;
+        }
+        throw new IllegalArgumentException(
+            "version may only be set to an int or a long but was [" + obj + "] with type [" + obj.getClass().getName() + "]"
+        );
+    }
+
+    public Object getRawVersion() {
+        return map.get(VERSION);
+    }
+
+    public MapBackedMetadata setVersion(Long version) {
+        return set(VERSION, version);
+    }
+
+    public void removeVersion() {
```

Comment:
When is it ok to remove version? It's odd having this and getRawVersion, which is different from the rest of the getters/setters.

Reply:
[IngestService](https://github.com/elastic/elasticsearch/blob/a4fdb0ba669cf473a3238faaa28f780d612ff65b/server/src/main/java/org/elasticsearch/ingest/IngestService.java#L908) expects this to be non-null and throws a 500 if it is null.